### PR TITLE
[bal-build-task]Integrate OpenAPI Client generation to Ballerina package build

### PR DESCRIPTION
**Pull request by lnash94**

### CODE REVIEW

**Comment by lnash94** [comment]
Please note that some tests will be added in the separate PR which introduces `bal openapi add` https://github.com/ballerina-platform/ballerina-library/issues/4932#issuecomment-1810005191

File: openapi-bal-task-plugin/src/main/resources/openapi-options-schema.json [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+      "type": "string"
+    },
+    "tags": {
+      "type": "string"
+    },
+    "nullable": {
+      "type": "string"
+    },
+    "mode": {
+      "type": "string"
```
Comment (TharmiganK) [inline openapi-bal-task-plugin/src/main/resources/openapi-options-schema.json:18]: If we are supporting enum type and there is an automatic validation, it is better to have that right?

File: openapi-bal-task-plugin/src/main/java/io/ballerina/openapi/bal/tool/OpenAPIBuildToolRunner.java [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+                    filter.setTags(List.of(values));
+                    break;
+                case OPERATIONS:
+                    filter.setOperations(List.of(values));
+                    break;
+                case NULLABLE:
+                    serviceMetaDataBuilder.withNullable(Arrays.asList(values).contains(TRUE));
+                    clientMetaDataBuilder.withNullable(Arrays.asList(values).contains(TRUE));
+                    break;
+                case IS_RESOURCE:
```
Comment (TharmiganK) [inline openapi-bal-task-plugin/src/main/java/io/ballerina/openapi/bal/tool/OpenAPIBuildToolRunner.java:203]: This should be `CLIENT_METHODS` right?

File: openapi-bal-task-plugin/src/main/java/module-info.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) 2023, WSO2 LLC. (http://www.wso2.com). All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (SachinAkash01) [reply openapi-bal-task-plugin/src/main/java/module-info.java:2]: Same case in [here](https://github.com/ballerina-platform/openapi-tools/pull/1569/files#:~:text=/*-,*%20Copyright%20(c)%202023%2C%20WSO2%20LLC.%20(http%3A//www.wso2.com).%20All%20Rights%20Reserved.,-*) as well.

File: openapi-bal-task-plugin/src/main/java/io/ballerina/openapi/bal/tool/OpenAPIBuildToolRunner.java [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+            TomlTableNode tomlTableNode = toolContext.optionsTable();
+            if (tomlTableNode == null) {
+                // Default generates client
+                generateClient(toolContext, codeGeneratorConfig);
+            } else {
+                Map<String, TopLevelNode> options = tomlTableNode.entries();
+                if (options.containsKey(MODE) && options.get(MODE).toString().contains(CLIENT)) {
+                        // Create client for the given OAS
+                        generateClient(toolContext, codeGeneratorConfig);
+                }
```
[the remark this answers is not in the document]
Comment (lnash94) [reply openapi-bal-task-plugin/src/main/java/io/ballerina/openapi/bal/tool/OpenAPIBuildToolRunner.java:135]: Yeah we need to provide a warning, updated with the warning :)

File: openapi-bal-task-plugin/src/main/java/io/ballerina/openapi/bal/tool/OpenAPIBuildToolRunner.java [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+            IOException, BallerinaOpenApiException {
+
+        Filter filter = new Filter();
+        OASClientConfig.Builder clientMetaDataBuilder = new OASClientConfig.Builder();
+        OASServiceMetadata.Builder serviceMetaDataBuilder = new OASServiceMetadata.Builder();
+        clientMetaDataBuilder.withOpenAPI(openAPI);
+        serviceMetaDataBuilder.withOpenAPI(openAPI);
+        TomlTableNode tomlTableNode = toolContext.optionsTable();
+        if (tomlTableNode == null) {
+            clientMetaDataBuilder.withFilters(filter);
```
[the remark this answers is not in the document]
Comment (lnash94) [reply openapi-bal-task-plugin/src/main/java/io/ballerina/openapi/bal/tool/OpenAPIBuildToolRunner.java]: Updated https://github.com/ballerina-platform/openapi-tools/pull/1569/commits/a0923a8802935dcf977e60fcb494f82579cb4812

File: openapi-bal-task-plugin/src/main/java/io/ballerina/openapi/bal/tool/OpenAPIBuildToolRunner.java (hunk 2 above)
Comment (lnash94) [reply openapi-bal-task-plugin/src/main/java/io/ballerina/openapi/bal/tool/OpenAPIBuildToolRunner.java]: will change to CLIENT_METHODS

File: openapi-bal-task-plugin/src/main/java/io/ballerina/openapi/bal/tool/OpenAPIBuildToolRunner.java [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+
+    @Override
+    public void executeTool(ToolContext toolContext) {
+
+        ImmutablePair<OASClientConfig, OASServiceMetadata> codeGeneratorConfig;
+        TomlNodeLocation location = toolContext.packageInstance().ballerinaToml().get().tomlAstNode().location();
+        try {
+            // Validate the OAS file whether we can handle within OpenAPI tool
+            if (!canHandle(toolContext)) {
+                return;
```
[the remark this answers is not in the document]
Comment (lnash94) [reply openapi-bal-task-plugin/src/main/java/io/ballerina/openapi/bal/tool/OpenAPIBuildToolRunner.java:107]: Yeah we need, handle via https://github.com/ballerina-platform/openapi-tools/pull/1569/commits/a0923a8802935dcf977e60fcb494f82579cb4812

File: openapi-bal-task-plugin/src/main/java/module-info.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (lnash94) [reply openapi-bal-task-plugin/src/main/java/module-info.java]: Updated :)

File: openapi-bal-task-plugin/src/main/resources/openapi-options-schema.json (hunk 1 above)
Comment (lnash94) [reply openapi-bal-task-plugin/src/main/resources/openapi-options-schema.json]: Tried with the enum value and got and JSON parser issue. FTM shall we go with string type? I will improve this with the enum once the JSON error is fixed.

**Review by TharmiganK — APPROVED** [review]

**Pull request merged**